2048 machine coding problem/GameBoardService.java

```diff
@@ -0,0 +1,226 @@
+package TwoZeroFourEight;
+
+import TwoZeroFourEight.GameBoard;
+import java.util.Random;
+
+public class GameBoardService extends GameBoard {
+    private int totalRows;
+    private int totalCols;
+    private static String final_board_val = "2048";
+    private Boolean isGameWon;
+    private Boolean isGameLost;
+    private Boolean isFirstAllocation;
+    private GameBoardServiceUtil util;
+
+    GameBoardService()
+    {
+        super();
+        this.totalRows = this.getTotalRows();
+        this.totalCols = this.getTotalCols();
+        this.isFirstAllocation = true;
+        this.isGameWon = false;
+        this.isGameLost = false;
+        this.util = new GameBoardServiceUtil();
+    }
+
+    public Boolean getGameWon() {
+        return isGameWon;
+    }
+
+    public Boolean getGameLost() {
+        return isGameLost;
+    }
+
+    public void setGameWon(Boolean gameWon) {
+        isGameWon = gameWon;
+    }
+
+    public void setGameLost(Boolean gameLost) {
+        isGameLost = gameLost;
+    }
+
+    void randomAllocateCellUtil(int position)
+    {
+        int positionCount = 0;
+        Boolean isAllocated = false;
+        for(int row = 0; row < this.getTotalRows(); row++)
+        {
+            for(int col = 0; col < this.getTotalCols(); col++)
+            {
+                if(this.getGameBoardCellValue(row,col)=="-"){
```

Comment:
Abstract this logic in board class. Do not expose the internal implementation (that board holds strings). The implementation should be easily changeable to numbers with empty being denoted by 0 without making any changes in the clients of the board class